message_verifier_test.go

```diff
@@ -0,0 +1,706 @@
+				},
+				Body: io.NopCloser(bytes.NewBuffer([]byte("test"))),
+			},
+			signature: `keyId="abc",algorithm="rsa-sha256",headers="(request-target) date host digest",signature="jFpHNN2jW4MTZnZGxm38qx6TqLJCeHEnKW7NhujN6VL7mFbZt6MZ1UcJS/YUYKLnFw1PBbvcOYjTarf2fCtIsCrHKCcYaSaK33L9HrcqoDH/ykOCd57TcNggrFCiqV4sOdjXgFj8Bi7fJbK0QCbIKv03YxBAUiUHArnTP6ANpRiAIyzQxGjGTFjgd2ENgU7Luf346xHdblbwecvJvabmN63mf9a/WFZ21JLERLH2PtAlcbmK43trIjnvHwGeSbi7Pd0xpiXXlhyG+iIdLqbszKjGvYd2cdK4nSfjJhhslA8KS8q5DVp0kJLZKfTbtAVI3mrmVyRAH0ZAa7SXqZQlcQ=="`,
+			requiredHeaders: []string{
+				"(request-target)",
+				"date",
+				"host",
+				"digest",
+			},
+			keyIDMetadataFn: getRSAPublicKey(0, nil),
+			errorExpected:   true,
+		},
+		{
+			title: "unsupported digest algorithm",
+			req: &http.Request{
+				Method: "POST",
+				URL: &url.URL{
+					Path: "/example",
+				},
+				Header: http.Header{
+					"Date":   {"Fri, 29 Jul 2022 13:23:35 GMT"},
+					"Host":   {"example.org"},
+					"Digest": {"SHA-1=n4bQgYhMfWWaL+qgxVrQFaO/TxsrC4Is0V1sFbDwCgg="},
+				},
+				Body: io.NopCloser(bytes.NewBuffer([]byte("test"))),
+			},
+			signature: `keyId="abc",algorithm="rsa-sha256",headers="(request-target) date host digest",signature="jFpHNN2jW4MTZnZGxm38qx6TqLJCeHEnKW7NhujN6VL7mFbZt6MZ1UcJS/YUYKLnFw1PBbvcOYjTarf2fCtIsCrHKCcYaSaK33L9HrcqoDH/ykOCd57TcNggrFCiqV4sOdjXgFj8Bi7fJbK0QCbIKv03YxBAUiUHArnTP6ANpRiAIyzQxGjGTFjgd2ENgU7Luf346xHdblbwecvJvabmN63mf9a/WFZ21JLERLH2PtAlcbmK43trIjnvHwGeSbi7Pd0xpiXXlhyG+iIdLqbszKjGvYd2cdK4nSfjJhhslA8KS8q5DVp0kJLZKfTbtAVI3mrmVyRAH0ZAa7SXqZQlcQ=="`,
+			requiredHeaders: []string{
+				"(request-target)",
+				"date",
+				"host",
+				"digest",
+			},
+			keyIDMetadataFn: getRSAPublicKey(0, nil),
+			errorExpected:   true,
+		},
+	}
+
+	for _, tc := range testCases {
+		// run test case with Authorization header
+		tc.req.Header.Set(string(Authorization), "Signature "+tc.signature)
+		t.Run(tc.title, tc.test)
+
+		// run test case with Signature header
+		tc.req.Header.Del(string(Authorization))
+		tc.req.Header.Set(string(Signature), tc.signature)
+		t.Run(tc.title, tc.test)
+	}
+}
+
+type extractSignatureTestCase struct {
+	title                    string
+	headers                  http.Header
+	expectedKeyID            string
+	expectedAlgorithm        string
+	expectedSignatureHeaders []string
+	expectedSignature        string
+	errorExpected            bool
+}
+
+func (tc *extractSignatureTestCase) test(t *testing.T) {
+	ma := &messageAttributes{
+		req: &http.Request{
+			Header: tc.headers,
+		},
+	}
+
+	err := ma.extractSignature()
+
+	if tc.errorExpected {
+		assert.Error(t, err)
+		return
+	}
+
+	assert.NoError(t, err)
+	assert.Equal(t, tc.expectedKeyID, ma.keyID)
+	assert.Equal(t, tc.expectedAlgorithm, ma.algorithm)
+	assert.Equal(t, tc.expectedSignatureHeaders, ma.signatureHeaders)
+	assert.Equal(t, tc.expectedSignature, ma.signature)
+}
+
+func Test_extractSignature(t *testing.T) {
+	testCases := []extractSignatureTestCase{
+		{
```

Comment:
It may be worth adding testcases that contain keys (so `keyId`, `algorithm` etc) in different order